File: src/main/java/mc/fbturrets/item/GunItem.java
package mc.fbturrets.item;

import mc.fbturrets.blocks.TurretGun;
import mc.fbturrets.blocks.TurretHolderBlockEntity;
import mc.fbturrets.main.FBTurrets;
import net.minecraft.block.entity.BlockEntity;
import net.minecraft.item.Item;
import net.minecraft.item.ItemUsageContext;
import net.minecraft.util.ActionResult;
import net.minecraft.util.Identifier;
import net.minecraft.util.registry.Registry;

import java.util.Objects;

public class GunItem extends Item {
    private final Identifier ID;
    
    public GunItem(Identifier itemID, Settings settings) {
        super(settings);
        ID = itemID;
    }
    
    @Override
    public ActionResult useOnBlock(ItemUsageContext context) {
        BlockEntity be = context.getWorld().getBlockEntity(context.getBlockPos());
        if (be != null && be.getType() == FBTurrets.TURRET_HOLDER_BLOCK_ENTITY) {
            if (((TurretHolderBlockEntity)be).gun.getID() != ID) {
                if (!Objects.requireNonNull(context.getPlayer()).isCreative()) {
                    ((TurretHolderBlockEntity)be).dropGun();
                    context.getPlayer().getMainHandStack().decrement(1);
                }
                ((TurretHolderBlockEntity)be).gun = (TurretGun)Registry.BLOCK.get(ID);
            }
            return ActionResult.SUCCESS;
        }
        return ActionResult.CONSUME;
    }
}

File: src/main/kotlin/mc/fbturrets/item/GunItem.kt
package mc.fbturrets.item

import mc.fbturrets.blocks.TurretGun
import mc.fbturrets.blocks.TurretHolderBlockEntity
import mc.fbturrets.main.FBTurrets
import net.minecraft.item.Item
import net.minecraft.item.ItemUsageContext
import net.minecraft.util.ActionResult
import net.minecraft.util.Identifier
import net.minecraft.util.registry.Registry
import java.util.*

class GunItem(private val ID: Identifier, settings: Settings?) : Item(settings) {
    override fun useOnBlock(context: ItemUsageContext): ActionResult {
        val be = context.world.getBlockEntity(context.blockPos)
        if (be != null && be.type === FBTurrets.TURRET_HOLDER_BLOCK_ENTITY) {
            val gun = (be as TurretHolderBlockEntity).gun
            if (gun != null && gun.iD !== ID) {
                if (!context.player!!.isCreative) {
                    be.dropGun()
                    context.player!!.mainHandStack.decrement(1)
                }
            }
            be.gun = Registry.BLOCK[ID] as TurretGun
            return ActionResult.SUCCESS
        }
        return ActionResult.CONSUME
    }

}